core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -502,16 +501,19 @@ private void doStop(CompletableFuture<Void> future,
                     }
 
                     workerShutdownFuture.addListener(unused5 -> {
-                        // If starts to shutdown before initializing serverChannels, completes the future
-                        // immediately.
-                        if (serverChannels.isEmpty()) {
+                        final Set<EventLoopGroup> bossGroups =
+                                ImmutableSet.copyOf(Server.this.serverChannels.keySet());
+
+                        // If started to shutdown before initializing a boss group,
+                        // complete the future immediately.
+                        if (bossGroups.isEmpty()) {
                             finishDoStop(future);
                             return;
                         }
+
                         // Shut down all boss groups and wait until they are terminated.
-                        final AtomicInteger remainingBossGroups = new AtomicInteger(serverChannels.size());
-                        serverChannels.forEach(ch -> {
-                            final EventLoopGroup bossGroup = ch.eventLoop().parent();
+                        final AtomicInteger remainingBossGroups = new AtomicInteger(bossGroups.size());
+                        bossGroups.forEach(bossGroup -> {
```

Comment:
I think this `bossGroup` is shadowing another one so maybe a different name is better.

Reply:
Hmm? What does it shadow?

Reply:
I couldn't find shadow too. Maybe GitHub diff fold lots of lines, and the 405 line makes you confused 😉

Reply:
Oops yeah that was it